llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -7792,6 +7582,29 @@ VectorizationFactor LoopVectorizationPlanner::computeBestVF() {
   VectorizationFactor LegacyVF = selectVectorizationFactor();
   VPlan &BestPlan = getPlanFor(BestFactor.Width);
 
+  // VPlan calculates register pressure from the plan, so it can come to
+  // different conclusions than the legacy cost model.
+  bool RegUsageDeterminedVF = false;
+  if (BestFactor.Width != LegacyVF.Width) {
+    SmallVector<ElementCount, 1> LegacyVFs = {LegacyVF.Width};
+    SmallVector<ElementCount, 1> VFs = {BestFactor.Width};
+
+    auto LegacyRUs =
+        calculateRegisterUsage(getPlanFor(LegacyVF.Width), LegacyVFs, TTI);
+    auto RUs = calculateRegisterUsage(BestPlan, VFs, TTI);
+
+    auto GetMaxUsage = [](SmallMapVector<unsigned, unsigned, 4> MaxLocalUsers) {
+      unsigned Max = 0;
+      for (auto Pair : MaxLocalUsers)
+        if (Pair.second > Max)
+          Max = Pair.second;
+      return Max;
```

Comment:
Although rather than creating a `GetMaxUsage`, you could use `max_element` for this instead, e.g.
```
auto CmpGT = [](std::pair<unsigned, unsigned> &A,
               std::pair<unsigned, unsigned> &B) {
  return A.second > B.second;
};
unsigned MaxLegacyRegUsage =
    max_element(LegacyRUs[0].MaxLocalUsers, CmpGT)->second;
unsigned MaxRegUsage = max_element(RUs[0].MaxLocalUsers, CmpGT)->second;
```